feat(cli): introduce the 'watch' command (#17240)
aws/aws-cdk: packages/aws-cdk/README.md

**⚠ Note #2**: This command is considered experimental,
and might have breaking changes in the future.

### `cdk watch`

> The `watch` command is similar to `deploy`,

Reviewer: ```suggestion
The `watch` command is similar to `deploy --hotswap`,
```

Author: It's explained below that `watch` implies `--hotswap`. Also, you can use `watch` without `--hotswap`, which is also explained below.

If you'd still like to to include `--hotswap` here, I won't oppose it, but I feel like it's a little misleading.

Reviewer: Okay fair enough

but instead of being a one-shot operation,
the command continuously monitors the files of the project,

> and triggers a `cdk deploy` whenever it detects any changes:

Reviewer: ```suggestion
and triggers a deployment whenever it detects any changes:
```

Author: 👍

```console

> $ cdk watch ExampleStack

Reviewer: What happens if you don't specify a stack name? Preferably, the first example should be the minimal default usage:

```suggestion
$ cdk watch
```

Author: The problem is that no argument means "deploy everything", which doesn't really make sense for "watch".

I've changed the example to `cdk watch DevelopmentStack`, let me know if that's better.

Reviewer: Why doesn't it make sense for watch to deploy everything?

Would that work?

Author: It does work, but `watch` is a development command - that's why it defaults `--hotswap` to `true`. For that reason, deploying _all_ stacks using it will be a rarity. In fact, for that reason, there might be a legitimate case for _not_ allowing all stacks to be deployed using `watch`, but not sure we want to go that far (it is useful for demos 😉).

Reviewer: I think we should still support deploying all stacks with watch. There could be a case for dev apps that consist of multiple stacks

Detected change to 'lambda-code/index.js' (type: change). Triggering 'cdk deploy'
ExampleStack: deploying...

 ✅  ExampleStack

^C
```

To end a `cdk watch` session, interrupt the process by pressing Ctrl+C.

What files are observed is determined by the `"watch"` setting in your `cdk.json` file.

> It has two sub-keys, `"include"` and `"exclude"`, each which can be either a single string, or an array of strings.

Reviewer: ```suggestion
It has two sub-keys, `"include"` and `"exclude"`, each of which can be either a single string, or an array of strings.
```

Author: 👍

Each entry is interpreted as a path relative to the location of the `cdk.json` file.
Globs, both `*` and `**`, are allowed to be used.
Example:

```json
{
  "app": "mvn -e -q compile exec:java",

>   "watch": {

Reviewer: I would also include a `build` setting here.

Author: Hmm. The `"build"` PR (https://github.com/aws/aws-cdk/pull/17176) is not merged yet. What if this gets merged before that PR is on `master`?

Reviewer: Good point, let's fast follow with an update

    "include": "src/main/**",
    "exclude": "target/*"
  }
}
```

> Note that `watch` by default uses hotswap deployments, and disables CloudFormation rollback.

Reviewer: Mention how this interacts with "build"

Author: Same response as https://github.com/aws/aws-cdk/pull/17240#discussion_r740621652.
